tools/js-sdk-release-tools/packages/typescript-codegen-breaking-change-detector/src/azure/diff/declaration-diff.ts

```diff
@@ -0,0 +1,433 @@
+// TODO: support class
+import {
+  InterfaceDeclaration,
+  Node,
+  ParameterDeclaration,
+  Signature,
+  SyntaxKind,
+  Symbol,
+  SymbolFlags,
+  FunctionDeclaration,
+  TypeNode,
+  TypeAliasDeclaration,
+  CallSignatureDeclaration,
+} from 'ts-morph';
+import {
+  DiffLocation,
+  DiffPair,
+  DiffReasons,
+  FindMappingCallSignature,
+  AssignDirection,
+  NameNode,
+} from '../common/types';
+import {
+  getCallableEntityParametersFromSymbol,
+  isMethodOrArrowFunction,
+  isPropertyArrowFunction,
+  isPropertyMethod,
+  isSameSignature,
+} from '../../utils/ast-utils';
+
+function findBreakingReasons(source: Node, target: Node): DiffReasons {
+  // Note: if return type node defined,
+  // it's a funtion/method/signature's return type node,
+  // return it, it will be used to compare later
+  // Otherwise, it's a non-funtion/method/signature node, return its type node
+  const getTypeNode = (node: Node): TypeNode => {
+    if (Node.isReturnTyped(node)) return node.getReturnTypeNodeOrThrow();
+    if (Node.isTyped(node)) return node.getTypeNodeOrThrow();
+    throw new Error(`Unsupported ${node.getKindName()} node: "${node.getText()}"`);
+  };
+  let breakingReasons = DiffReasons.None;
+
+  const targetTypeNode = getTypeNode(target);
+  const sourceTypeNode = getTypeNode(source);
+
+  // check if concrete type -> any. e.g. string -> any
+  const isConcretTypeToAny = canConvertConcretTypeToAny(targetTypeNode?.getKind(), sourceTypeNode?.getKind());
+  if (isConcretTypeToAny) breakingReasons |= DiffReasons.TypeChanged;
+
+  // check type predicates
+  if (
+    targetTypeNode &&
+    sourceTypeNode &&
+    targetTypeNode.isKind(SyntaxKind.TypePredicate) &&
+    sourceTypeNode.isKind(SyntaxKind.TypePredicate)
+  ) {
+    const getTypeName = (node: TypeNode) => node.asKindOrThrow(SyntaxKind.TypePredicate).getTypeNodeOrThrow().getText();
+    if (getTypeName(targetTypeNode) !== getTypeName(sourceTypeNode)) breakingReasons |= DiffReasons.TypeChanged;
+  }
+
+  // check type
+  const assignable = sourceTypeNode.getType().isAssignableTo(targetTypeNode.getType());
+  if (!assignable) breakingReasons |= DiffReasons.TypeChanged;
+
+  // check required -> optional
+  const isOptional = (node: Node) => node.getSymbolOrThrow().isOptional();
+  const incompatibleOptional = isOptional(target) && !isOptional(source);
+  if (incompatibleOptional) breakingReasons |= DiffReasons.RequiredToOptional;
```

Comment:
Can we differentiate if it's request or response? I think if it's request, changing from required to optional would be a non-breaking.

Reply:
we don't need to tell request/response here, since this function is general, the assignment is always from source to target. so, for input model, we call `findBreakingReasons(old, new)`, for output model we call `findBreakingReasons(new, old)`